modules/core/src/test/java/org/apache/ignite/internal/processors/metastorage/persistence/DistributedMetaStoragePersistentTest.java

```diff
@@ -543,6 +571,40 @@ public void testConflictingData() throws Exception {
         );
     }
 
+    /** */
+    @Test
+    public void testReplayOnJoinSkipsHistoryItemsOptimization() throws Exception {
+        IgniteEx igniteEx = startGrid(0);
+
+        igniteEx.cluster().baselineAutoAdjustEnabled(false);
+
+        startGrid(1);
+
+        grid(0).cluster().active(true);
+
+        metastorage(0).write("key1", "value1");
+
+        stopGrid(1);
+
+        JdkMarshaller marshaller = GridTestUtils.getFieldValue(metastorage(0), "marshaller");
+        GridTestUtils.invoke(
+            metastorage(0),
+            "completeWrite",
+            new DistributedMetaStorageHistoryItem("key1", marshaller.marshal("value1")),
+            false
+        );
+
+        startGrid(1);
+
+        metastorage(0).write("key2", "value2");
+
+        stopGrid(0);
+        stopGrid(1);
+
+        startGrid(0);
+        startGrid(1);
+    }
```

Comment:
I would ask you to write in a comment what is expected in the end, if it is not too much trouble, it is not quite obvious to me.

Reply:
Sure, I'll do it